### DISCUSSION TITLE
Syntax error in python script when python version 2 is used

### DISCUSSION THREAD

**Discussion by yaaminiv** [RobertsLab/resources#1605] [Q&A] [answered]
I am using the following script to get unreviewed and reviewed GOterm information from SwissProt, modified from instructions on [this page](https://www.uniprot.org/help/api_queries) under "2.2 Large number of results: use pagination":

```
#!/usr/bin/env python

import re
import requests
from requests.adapters import HTTPAdapter, Retry

re_next_link = re.compile(r'<(.+)>; rel="next"')
retries = Retry(total=5, backoff_factor=0.25, status_forcelist=[500, 502, 503, 504])
session = requests.Session()
session.mount("https://", HTTPAdapter(max_retries=retries))

def get_next_link(headers):
    if "Link" in headers:
        match = re_next_link.match(headers["Link"])
        if match:
            return match.group(1)

def get_batch(batch_url):
    while batch_url:
        response = session.get(batch_url)
        response.raise_for_status()
        total = response.headers["x-total-results"]
        yield response, total
        batch_url = get_next_link(response.headers)

url = 'https://rest.uniprot.org/uniprotkb/search?compressed=true&fields=accession%2Creviewed%2Cid%2Cprotein_name%2Cgene_names%2Corganism_name%2Clength%2Cgo%2Cgo_p%2Cgo_c%2Cgo_f%2Cgo_id&format=tsv&query=%28%2A%29&size=500'
progress = 0
with open('uniprot-all-20230203.tsv', 'w') as f:
    for batch, total in get_batch(url):
        lines = batch.text.splitlines()
        if not progress:
            print(lines[0], file=f)
        for line in lines[1:]:
            print(line, file=f)
        progress += len(lines[1:])
        print(f'{progress} / {total}')
```

I can run this code on my local machine (python v.3.7.4 ) but it takes ~10 days to complete. I tried running this on our cluster, but the output ended up being garbled. That is not an issue when I run it on a local machine, so I thought I'd try using raven (Grace ran the script for me since I don't have VPN access). I got the following error from raven: 

<img width="469" alt="Screenshot 2023-03-07 at 3 43 09 PM" src="https://user-images.githubusercontent.com/22335838/223547826-d2cde95d-be57-4556-bd28-705d2780c997.png">

I got the same error on ostrich. Both raven and ostrich have python v.2. Is there a reason why the syntax in this script is not working for machines with python v.2? Is there something else going on preventing me from running my script on raven or ostrich?

**Comment by kubu4** [comment, marked as answer]
>  Is there a reason why the syntax in this script is not working for machines with python v.2?

The instructions indicate that Python code is written to use Python v3, so I wouldn't expect it to work with a Python v2.x version.

Run conda/miniconda to use an updated Python version. You might have to install it.

**Reply by kubu4** [reply]
Actually, you _won't_ need to install on Raven!

See these instructions:

https://github.com/sr320/ceabigr/issues/35#issuecomment-1047919208

**Comment by kubu4** [comment]
> but it takes ~10 days to complete.

Knowing this, I'd highly recommend you start a `screen` session before executing the script. This will allow you to "detach" the `screen` session so that the script will still run, even if you turn off your computer or lose connection to Raven.

**Reply by yaaminiv** [reply]
yep, planned on starting a `tmux` session to run the script
